src/main/java/com/puppycrawl/tools/checkstyle/DetailAstImpl.java

```diff
@@ -138,21 +138,15 @@ else if (parent != null) {
      * @param ast DetailAST object.
      */
     public void addNextSibling(DetailAST ast) {
-        clearBranchTokenTypes();
-        clearChildCountCache(parent);
-        if (ast != null) {
-            // parent is set in setNextSibling
-            final DetailAstImpl sibling = nextSibling;
-            final DetailAstImpl astImpl = (DetailAstImpl) ast;
+        // parent is set in setNextSibling
+        final DetailAstImpl sibling = nextSibling;
+        final DetailAstImpl astImpl = (DetailAstImpl) ast;
 
-            if (sibling != null) {
-                astImpl.setNextSibling(sibling);
-                sibling.previousSibling = astImpl;
```

Comment:
In the `astImpl.setNextSibling` we assign `sibling.previousSibling = this`, after that we have `sibling.previousSibling = astImpl`. It doesn't look like `this` is the same link as `astImpl`. Why did you remove it?

Reply:
For Explanation, I will take an example. 

Let's suppose,
We have some structure where we have nodes `X, Y, Z, W`  where `X` is the parent node `Y` is the first child and `Z` is a sibling of `Y`
```
 X
 |
  ----Y-----Z

```

Now we have to add the `W` node as nextSibiling of `Y` 
```
 X
 |
  ----Y-----W----Z

```
 

Know the current code is 

```
    public void addNextSibling(DetailAST ast) {
        clearBranchTokenTypes();
        clearChildCountCache(parent);
        if (ast != null) {
            // parent is set in setNextSibling
            final DetailAstImpl sibling = nextSibling;
            final DetailAstImpl astImpl = (DetailAstImpl) ast;

            if (sibling != null) {
                astImpl.setNextSibling(sibling);
                sibling.previousSibling = astImpl;
            }

            astImpl.previousSibling = this;
            setNextSibling(astImpl);
        }
    }
```

know to add the  `W` node as the sibling of `Y` in the existing structure. 
we call the method of object `Y` addNextSibling something like `Y.addNextSibling(W)` 

Then the sibling is assigned to node nextSibiling Which will be our Node `Z`
and 
astImpl will be our Node `W` 
```
 final DetailAstImpl sibling = nextSibling;
 final DetailAstImpl astImpl = (DetailAstImpl) ast;
```

Know in `if` statement  we are calling setNextSibiling where astImpl is our last means object `W` and sibling is object `Z`
```
astImpl.setNextSibling(sibling);
```

know we directly move to `setNextSibling`
```
    public void setNextSibling(DetailAST nextSibling) {
        clearBranchTokenTypes();
        clearChildCountCache(parent);
        this.nextSibling = (DetailAstImpl) nextSibling;
        if (nextSibling != null && parent != null) {
            ((DetailAstImpl) nextSibling).setParent(parent);
        }
        if (nextSibling != null) {
            ((DetailAstImpl) nextSibling).previousSibling = this;
        }
    }
```
  
here first `this.nextSibling = (DetailAstImpl) nextSibling;` here `this` keyword will point the object `W`. (The reason is https://docs.oracle.com/javase/tutorial/java/javaOO/thiskey.html#:~:text=this%20is%20a%20reference%20to%20the%20current%20object%20%E2%80%94%20the%20object%20whose%20method%20or%20constructor%20is%20being%20called here W 's object's method is called so this right know is W) and nextSibiling is Z.
```
W --> Z
```

know if we look directly at 
```
((DetailAstImpl) nextSibling).previousSibling = this;
```
here our nextSibiling is Z and this will refer to `W` means know 
```
W <----> Z 
```

know in the addNextsibiling the mutated one line `sibling.previousSibling = astImpl;` here if we recall sibling is `Z` and astIMpl is `W` 
know we again doing the stuff 
```
W <----  Z 
```

which is already done in setNextSibiling

Reply:
@rdiachenko  I hope I make the doubt clear :)